subsys/random/rand32_entropy_device.c

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <atomic.h>
+#include <entropy.h>
+
+static atomic_t entropy_driver;
+
+u32_t sys_rand32_get(void)
+{
+	struct device *dev = (struct device *)atomic_get(&entropy_driver);
+	u32_t random_num;
+	int ret;
+
+	if (unlikely(!dev)) {
+		/* Only one entropy device exists, so this is safe even
+		 * if the whole operation isn't atomic.
+		 */
+		dev = device_get_binding(CONFIG_ENTROPY_NAME);
+		atomic_set(&entropy_driver, (atomic_t)(uintptr_t)dev);
+	}
+
+	ret = entropy_get_entropy(dev, (u8_t *)buffer, sizeof(random_num))
+	if (unlikely(ret < 0)) {
```

Comment:
I realize this API doesn't have any way of communicating failure, but done this way, doesn't this mean that some kind of device failure would result in no randomness being returned? This is going to be bad even for things that don't want strong randomness (TCP delays and such)? Some thoughts:
* log the failure at least once? I'm not sure how common it is to have logging in Zephyr devices.
* Fall back to a pseudo-random sequence?
Or is this so unlikely, that in reality, none of the devices should ever fail.

Also, what are the requirements on sys_rand32_get? And have we defined what kind of quality entropy_get_entropy should return? Looking at the drivers, it seems that some try to be fairly good PRNGs and other make no claims of the sort. I'm concerned about some kind of entropy bias creating patterns in the output of this.

Admittedly, this isn't any different than the version before this patch, so this should probably go in this way, but this is something to think about.

Reply:
The purpose of this patch isn't to fix the random story in Zephyr; it's just a stepping stone to have it fixed. So, yes, it's not worse than what we currently have. It's more of a cleanup.

The random device might fail (some HWPRNGs will signal error if their entropy tank is empty, and this might happen in very early boot situations), so this is a possibility that has to be handled. Returning the CPU cycle counter in these situations might be better than nothing; will do that here.

Reply:
suggest maybe putting an __ASSERT for this case